### ISSUE TITLE
Remove single point of failure in the deployment process

### ISSUE THREAD

**Issue by npepinpe** [zeebe-io/zeebe#4765]
**Description**

Deployments currently work by always executing the deployment command on partition 1, which is then in charge of distributing that deployment to other partitions. This is fairly robust in the sense that we will retry the distribution until all partitions acknowledge that they have deployed the process definition.

There's a few issues with this approach, the main one being that partition 1 (e.g. the deployment partition) becomes a single point of failure in the process. If partition 1 has no leader, for example, no deployments are possible. If partition 1 receives it, then fails over or crashes, then it delays the distribution to the other partitions _although_ from a user point of view the deployment went through.

**Comment by saig0** [comment]
A bit background: 
The main reason for deploying only on one partition is the versioning of the workflows. Each deployed workflow has a version that is incremented for every new workflow with the same BPMN process id. 
If the deployment is not coordinated then it may be hard to keep the versioning consistent (in case of multiple deployments in a short time).

**Comment by Zelldon** [comment]
In addition to what @saig0 it is also that you need a master which keeps track of it. Otherwise you can't be sure that it is deployed on all partitions, even on restart. If you do it on the gateway what happens when the gateway fails? Currently the gateway is "stateless".

**Comment by npepinpe** [comment]
Yes, sorry, I should've mentioned why we're doing it this way already :slightly_smiling_face: 

I mainly created it to challenge this again, as there is a clear single point of failure which I would hope to avoid in the long run.

**Comment by Zelldon** [comment]
I mean normally we expect that we eventually have a leader again right?

**Comment by npepinpe** [comment]
Yes, though as we've seen this isn't always that fast :sweat_smile: 

Maybe the crux becomes more how we report the deployment to users - I know it's been a source of confusion, and users expect that when they get a successful response they can create instances (which isn't always true).

**Comment by Zelldon** [comment]
I still not get why this is related to deployment process? I mean with message start events it also works right?

**Comment by npepinpe** [comment]
I'm not sure - I guess because if you create them on all partitions then you would get too many started at the same time, which is also counter intuitive, and timers have no real ID to hash to distribute across partitions?

**Comment by Zelldon** [comment]
Why ? Each deployment has a key? If you mod it with partition then you know which partition is in charge to start the timer?

**Comment by npepinpe** [comment]
We want to cancel the previous timer as well when deploying a new version, no? So I guess we could use the bpmnProcessId or something like that instead?

**Comment by Zelldon** [comment]
yeah probably

**Comment by npepinpe** [comment]
Superseded by #6488
